### PR TITLE
Bug 1927168: pkg/cvo/sync_worker: Increment Attempt on failed reconciliation

### PR DESCRIPTION
**Pull request by wking**
Avoid situations [like][1]:

```console
$ grep 'Running sync.*in state\|Result of work' cvo.log | tail -n4
I0517 11:30:37.342895       1 sync_worker.go:549] Running sync 4.8.0-0.nightly-2021-05-15-141455 (force=false) on generation 1 in state Reconciling at attempt 0
I0517 11:36:19.258793       1 task_graph.go:555] Result of work: [Could not update servicemonitor "openshift-machine-api/cluster-autoscaler-operator" (255 of 676): the server does not recognize this resource, check extension API servers Could not update servicemonitor "openshift-kube-apiserver-operator/kube-apiserver-operator" (627 of 676): the server does not recognize this resource, check extension API servers update context deadline exceeded at 28 of 676 update context deadline exceeded at 28 of 676]
I0517 11:39:38.268693       1 sync_worker.go:549] Running sync 4.8.0-0.nightly-2021-05-15-141455 (force=false) on generation 1 in state Reconciling at attempt 0
I0517 11:45:20.182573       1 task_graph.go:555] Result of work: [Could not update servicemonitor "openshift-machine-api/cluster-autoscaler-operator" (255 of 676): the server does not recognize this resource, check extension API servers Could not update servicemonitor "openshift-kube-apiserver-operator/kube-apiserver-operator" (627 of 676): the server does not recognize this resource, check extension API servers]
```

where we are failing the reconcile, but not incrementing `Attempt`.  Because we rely on `Attempt` moving to drive changes in the `ShiftOrder` and `PermuteOrder` calls in `SyncWorker.apply`.  This fixes a bug we've had since the permuation logic landed in 1f95ccfa05 (#166).

As a side benefit, this change will no longer call `InitCOUpdateStartTime` in `SyncWorker.syncOnce` after a failed reconciliation cycle, so if a ClusterOperator goes `Degraded` during reconciliation, we will now complain about it after the `UpdateEffectFailAfterInterval` threshold elapses.

I'm a bit suspicious about `steps := 8` capping `ShiftOrder`, because we have some manifest task-nodes that have a good deal more than that.  But blindly increasing it would reduce the amount of permuting `PermuteOrder` was doing.  We might have to do something more drastic for reconciliation mode, like collecting manifest failures and eventually failing the sync cycle, but not aborting it early to report that failure.  But one step at a time; I'm not going to overhaul all of that in this commit.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1927168#c10

### CODE REVIEW

**Comment by wking** [comment]
Cluster-bot cluster, via `launch openshift/cluster-version-operator#569`:

```console
$ oc get clusterversion -o jsonpath='{.status.desired.version}{"\n"}' version
4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest
$ oc delete crd servicemonitors.monitoring.coreos.com
customresourcedefinition.apiextensions.k8s.io "servicemonitors.monitoring.coreos.com" deleted
$ oc -n openshift-cluster-version get pods
NAME                                       READY   STATUS    RESTARTS   AGE
cluster-version-operator-dbcdd99d4-9hv4f   1/1     Running   0          39m
$ oc -n openshift-cluster-version logs cluster-version-operator-dbcdd99d4-9hv4f | head -n1
I0519 17:10:31.708389       1 start.go:21] ClusterVersionOperator v1.0.0-525-g8650e476-dirty
$ oc -n openshift-cluster-version logs cluster-version-operator-dbcdd99d4-9hv4f | grep 'Running sync.*in state\|Result of work'
I0519 17:14:12.784365       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Initializing at attempt 0
I0519 17:17:03.746322       1 task_graph.go:555] Result of work: [Cluster operator authentication is not available Cluster operator monitoring is not available Cluster operator console is not available]
I0519 17:17:26.598895       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Initializing at attempt 1
I0519 17:20:17.556997       1 task_graph.go:555] Result of work: [Cluster operator authentication is not available Cluster operator console is not available]
I0519 17:21:06.994799       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Initializing at attempt 2
I0519 17:23:57.952505       1 task_graph.go:555] Result of work: [Cluster operator authentication is not available]
I0519 17:25:26.187900       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Initializing at attempt 3
I0519 17:28:17.145563       1 task_graph.go:555] Result of work: [Cluster operator authentication is not available]
I0519 17:31:18.530222       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Initializing at attempt 4
I0519 17:31:39.119608       1 task_graph.go:555] Result of work: []
I0519 17:34:30.076980       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Reconciling at attempt 0
I0519 17:34:57.449939       1 task_graph.go:555] Result of work: []
I0519 17:37:48.406792       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Reconciling at attempt 0
I0519 17:38:15.812221       1 task_graph.go:555] Result of work: []
I0519 17:41:06.769355       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Reconciling at attempt 0
I0519 17:41:34.154324       1 task_graph.go:555] Result of work: []
I0519 17:44:25.111269       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Reconciling at attempt 0
I0519 17:44:52.484736       1 task_graph.go:555] Result of work: []
I0519 17:47:43.441953       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Reconciling at attempt 0
I0519 17:53:25.355226       1 task_graph.go:555] Result of work: [Could not update servicemonitor "openshift-cluster-node-tuning-operator/node-tuning-operator" (365 of 694): the server does not recognize this resource, check extension API servers Could not update servicemonitor "openshift-kube-apiserver-operator/kube-apiserver-operator" (645 of 694): the server does not recognize this resource, check extension API servers]
I0519 17:56:24.502884       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Reconciling at attempt 1
I0519 18:02:06.416771       1 task_graph.go:555] Result of work: [Could not update servicemonitor "openshift-machine-api/cluster-autoscaler-operator" (273 of 694): the server does not recognize this resource, check extension API servers Could not update servicemonitor "openshift-console-operator/console-operator" (631 of 694): the server does not recognize this resource, check extension API servers update context deadline exceeded at 71 of 694]
I0519 18:05:03.089286       1 sync_worker.go:541] Running sync 4.8.0-0.ci.test-2021-05-19-163242-ci-ln-q0ht5wt-latest (force=false) on generation 1 in state Reconciling at attempt 2
```

hooray :)

File: pkg/cvo/sync_worker.go [hunk 1]
```diff
@@ -349,7 +349,8 @@ func (w *SyncWorker) Start(ctx context.Context, maxWorkers int, cvoOptrName stri
 					}
 				}
 				next = time.After(wait.Jitter(interval, 0.2))
-
+				work.Completed = 0
+				work.Attempt++
```
Comment (LalatenduMohanty) [inline pkg/cvo/sync_worker.go:353]: :+1:

File: pkg/cvo/sync_worker.go [hunk 2]
```diff
@@ -358,6 +359,7 @@ func (w *SyncWorker) Start(ctx context.Context, maxWorkers int, cvoOptrName stri
 			}
 
 			work.Completed++
+			work.Attempt = 0
```
Comment (LalatenduMohanty) [inline pkg/cvo/sync_worker.go:362]: :+1:

File: pkg/cvo/sync_worker.go [hunk 3]
```diff
@@ -416,20 +418,10 @@ func (w *SyncWorker) calculateNext(work *SyncWork) bool {
...
-	}
-	// always clear the completed variable if we are not reconciling
-	if work.State != payload.ReconcilingPayload {
-		work.Completed = 0
-	}
-
-	// track how many times we have tried the current payload in the current
-	// state
-	if changed || w.work.State != work.State {
 		work.Attempt = 0
```
Comment (LalatenduMohanty) [inline pkg/cvo/sync_worker.go:424]: It is not clear why we need to make Attempt= 0 here?  Because it means `equalSyncWork returned false`

**Review by LalatenduMohanty — CHANGES_REQUESTED** [review]

File: pkg/cvo/sync_worker.go (hunk 3 above)
Comment (wking) [reply pkg/cvo/sync_worker.go:424]: We want to reset the count, because we are about to begin our first attempt at the new target.

**Review by LalatenduMohanty — APPROVED** [review]
/lgtm

**Comment by wking** [comment]
Azure is [having a tough day][1]:

image

but that's orthogonal to this PR.

/override ci/prow/e2e-agnostic-upgrade 

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_cluster-version-operator/569/pull-ci-openshift-cluster-version-operator-master-e2e-agnostic-upgrade/1395103918928695296

**Pull request merged**